game-core/src/main/java/games/strategy/triplea/ui/CommentPanel.java

```diff
@@ -98,10 +96,8 @@ private void createComponents() {
     save.setMargin(inset);
     save.setFocusable(false);
     for (final PlayerID playerId : data.getPlayerList().getPlayers()) {
-      final CompletableFuture<?> future = CompletableFuture
-          .supplyAsync(() -> frame.getUiContext().getFlagImageFactory().getSmallFlag(playerId))
-          .thenAccept(image -> SwingUtilities.invokeLater(() -> iconMap.put(playerId, new ImageIcon(image))));
-      CompletableFutureUtils.logExceptionWhenComplete(future, "Failed to load small flag icon for " + playerId);
+      Optional.ofNullable(frame.getUiContext().getFlagImageFactory().getSmallFlag(playerId))
+          .ifPresent(image -> iconMap.put(playerId, new ImageIcon(image)));
```

Comment:
The reason this asynchronous computation was introduced was in order to move this potential "expensive IO operation" off the EDT

Reply:
Agreed, but what speaks against a simple mutex to guarantee exclusive access?
Preferably using a normal thread to not block the thread pool unnecessarily

Reply:
Such a change would affect lots of code that uses `ImageFactory`.  That would require a lot of testing to verify no new regressions or performance issues were introduced.  I simply want to get rid of the 4-5 NPEs I'm experiencing each week due to this bug. :smile:

Also, a single mutex controlling access to `ImageFactory#getImage()` would kind of defeat the purpose of the optimization from #3383.  Why launch one thread per player all competing for the same mutex?  Might as well just load all images sequentially on a single background thread given how the code in `getImage()` is currently written (because `ImageIO#read()` would be protected by that mutex).  (I understand that the time to initially display the game window is reduced in either case; I'm just pointing out that it might actually take longer to load and display the images by using more threads than necessary.)

I would propose that a more fundamental change needs to be made to `ImageFactory` to support asynchronous loading (and caching) of images in a truly concurrent fashion.  A dedicated method or methods for such support would be a nice addition (e.g. `ImageFactory#loadImageAsync(String key, Consumer<Image>)`, possibly using `SwingWorker` internally).  That should be explored in a dedicated PR.